### PR TITLE
5397 callout banner

### PR DESCRIPTION
**Pull request by lzach83**
## Link to Issue
Closes: #5397 

## Description of Changes
- Implemented a call out banner on threads to show gated threads

## "How We Fixed It"

- Gathered all of the groups with the current topic and rendered the banner based on if the current user is a member.

## Test Plan
- Make a group 
- Test that the banner shows for those topic(s)

## Other Considerations

image

### CODE REVIEW

**Comment by lzach83** [comment]
To speed up the review, files of concern here are: 

`ViewThreadPage`
`CWBanner`
`CWBanner.scss`

All other file changes are related to linting

File: packages/commonwealth/client/scripts/views/pages/Community/Groups/common/GroupForm/GroupForm.tsx [hunk 1]
```diff
@@ -96,12 +97,21 @@ const GroupForm = ({
...
+    chainId: app.activeChainId()
   });
+
+  const { data: groups = [] } = useFetchGroupsQuery({
+    chainId: app.activeChainId()
+  });
+
+  const takenGroupNames = groups.map(({ name }) => name.toLowerCase());
+
+  const [isNameTaken, setIsNameTaken] = useState(false);
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/Community/Groups/common/GroupForm/GroupForm.tsx:113]: I think this should be in the group name uniqueness pr.

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWBanner.tsx [hunk 2]
```diff
@@ -41,6 +42,7 @@ interface CWBannerProps {
   buttons?: ButtonProps[];
   className?: string;
   onClose: () => void;
+  gatedGroups?: Array<{ id: number; name: string }>;
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWBanner.tsx:45]: can we make this a generic react node like `footer: ReactNode` and add the gated tags from wherever the component gets used?

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 3]
```diff
@@ -477,6 +520,22 @@ const ViewThreadPage = ({ identifier }: ViewThreadPageProps) => {
                       canComment={canComment}
                       shouldFocusEditor={shouldFocusCommentEditor}
                     />
+                    {!hideGatingBanner && gatedGroups.length > 0 && (
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:523]: Unrealted: based on the memberships we would have to show/hide this which will happen in #5465

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 4]
```diff
@@ -477,6 +520,22 @@ const ViewThreadPage = ({ identifier }: ViewThreadPageProps) => {
...
+                      <CWBanner
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:524]: We should add this based on the `featureflags.gatingEnabled` feature flag

**Review by mzparacha — COMMENTED** [review]
Small suggestions above. Thanks for the working on this 🙏.

File: packages/commonwealth/client/scripts/views/pages/Community/Groups/common/GroupForm/GroupForm.tsx (hunk 1 above)
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/pages/Community/Groups/common/GroupForm/GroupForm.tsx:113]: It is done in that PR as well: https://github.com/hicommonwealth/commonwealth/pull/5454/files#diff-4c2f41ca45be294fdb637836edeaf7b058571665e4cd00325fa332d288c18ccf

**Comment by lzach83** [comment]
Ready for review again.

**Comment by mzparacha** [comment]
I get some terminal/dev process errors

image

**Comment by lzach83** [comment]
> I get some terminal/dev process errors
> 
> image image

Failed merge conflict

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 5]
```diff
@@ -477,6 +521,34 @@ const ViewThreadPage = ({ identifier }: ViewThreadPageProps) => {
                       canComment={canComment}
                       shouldFocusEditor={shouldFocusCommentEditor}
                     />
+                    {featureFlags.gatingEnabled && gatedGroups.length > 0 && (
+                      <CWBanner
+                        title="This topic is gated"
+                        body="Only members within the following group(s) can interact with this topic:"
+                        type="info"
+                        footer={
+                          gatedGroups && (
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:530]:
```diff
-- gatedGroups && (
```

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 6]
```diff
@@ -477,6 +521,34 @@ const ViewThreadPage = ({ identifier }: ViewThreadPageProps) => {
...
+                              ))}
+                            </div>
+                          )
+                        }
+                        buttons={[
+                          {
+                            label: 'See all groups',
+                            onClick: () => navigate('/members?tab=groups'),
+                          },
+                          { label: 'Learn more about gating' },
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:547]: i am not sure, but shouldn't we redirect somewhere if they press on "lear more about gating" button?

If no, plz feel free to resolve this comment. Thanks.

Comment (lzach83) [reply packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:547]: Unsure if that screen is created and it's not indicated in figma. 

CC: @zakhap any input here?

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 7]
```diff
@@ -153,6 +165,38 @@ const ViewThreadPage = ({ identifier }: ViewThreadPageProps) => {
...
+        const isMemberOfAnyGroup = groupsWithTopic.some((community) =>
+          community.members.some((id) => addressTopicMemberIds.includes(id))
+        );
+
+        // You can use the 'isMemberOfAnyGroup' variable as needed
+        setHideGatingBanner(isMemberOfAnyGroup);
+      }
+    }
+  }, [groups, thread?.topic]);
+
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:199]: i would suggest removing this and the `gatedGroups` var from the state (we actually don't need these extra checks)

```diff
// find if the current topic is gated
const foundGatedTopic = groups.find(x => x.topic.find(y => y.id === topicId))

// Check if the user is a member of any of the groups with the current topic
if(foundGatedTopic) {
  const isMemberOfGatedTopic = foundGatedTopic.members.find(x => x?.address?.address === app.user.activeAccount.address)
}
```

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 8]
```diff
@@ -98,12 +101,21 @@ const ViewThreadPage = ({ identifier }: ViewThreadPageProps) => {
...
+  const [hideGatingBanner, setHideGatingBanner] = useState(false);
+  const [gatedGroups, setGatedGroups] = useState([]);
+
   const { isBannerVisible, handleCloseBanner } = useJoinCommunityBanner();
   const { handleJoinCommunity, JoinCommunityModals } = useJoinCommunity();
   const { activeAccount: hasJoinedCommunity } = useUserActiveAccount();
   const [searchParams] = useSearchParams();
   const shouldFocusCommentEditor = !!searchParams.get('focusEditor');
 
+  const { data: groups } = useFetchGroupsQuery({
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:113]:
```diff
-- const { data: groups } = useFetchGroupsQuery({
++ const { data: groups = [] } = useFetchGroupsQuery({
```

fallback is always nice :)

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 9]
```diff
@@ -477,6 +521,34 @@ const ViewThreadPage = ({ identifier }: ViewThreadPageProps) => {
                       canComment={canComment}
                       shouldFocusEditor={shouldFocusCommentEditor}
                     />
+                    {featureFlags.gatingEnabled && gatedGroups.length > 0 && (
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:524]: i suggest to remove `gatedGroups.length > 0` and [check if the current topic is gated](https://github.com/hicommonwealth/commonwealth/pull/5481/files#r1377921094), and only then show this banner.

**Review by mzparacha — CHANGES_REQUESTED** [review]
Some suggestions, thanks for the hard work on this 🙏

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx (hunk 8 above)
Comment (lzach83) [reply packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:113]: Great catch

**Review by mzparacha — APPROVED** [review]

File: packages/commonwealth/client/scripts/views/pages/Community/Groups/common/GroupForm/GroupForm.tsx (hunk 1 above)
Comment (rbennettcw) [reply packages/commonwealth/client/scripts/views/pages/Community/Groups/common/GroupForm/GroupForm.tsx:113]: I know this is already done, but for future reference, I think that this kind of state could be simpler with `useMemo` instead– having it check if `takenGroupNames` contains the new group name.

**Review by rbennettcw — APPROVED** [review]

**Pull request merged**